src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp

```diff
@@ -0,0 +1,191 @@
+// found in the LICENSE file.
+
+#include <gmock/gmock.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/support/channel_arguments.h>
+#include <gtest/gtest.h>
+
+#include <thread>
+
+#include "CaptureFile/CaptureFileOutputStream.h"
+#include "OrbitBase/Logging.h"
+#include "ProducerEventProcessor/UploaderClientCaptureEventCollector.h"
+#include "capture.pb.h"
+#include "services.grpc.pb.h"
+
+using orbit_grpc_protos::ClientCaptureEvent;
+
+namespace orbit_producer_event_processor {
+
+namespace {
+
+static constexpr const char* kAnswerString = "Test Interned String";
+static constexpr uint64_t kAnswerKey = 42;
+
+ClientCaptureEvent CreateInternedStringCaptureEvent(uint64_t key, const std::string& str) {
+  ClientCaptureEvent event;
+  event.mutable_interned_string()->set_key(key);
+  event.mutable_interned_string()->set_intern(str);
+  return event;
+}
+
+ClientCaptureEvent CreateCaptureFinishedEvent() {
+  ClientCaptureEvent event;
+  event.mutable_capture_finished()->set_status(orbit_grpc_protos::CaptureFinished::kSuccessful);
+  return event;
+}
+
+class UploaderClientCaptureEventCollectorTest : public testing::Test {
+ protected:
+  // Set `early_stop_event_count` to be >= 2 if we want to call `Stop()` early with producing only
+  // `early_stop_event_count` events. Here we require ">= 2" as an early stop will create at
+  // least two capture events: one `InternedStringCaptureEvent` created when stop_called_ becomes
+  // true, and one `CaptureFinishedEvent`.
+  void Start(size_t event_count, uint64_t produce_event_every_ms, uint64_t upload_events_every_ms,
+             size_t early_stop_event_count = 0) {
+    event_count_ = event_count;
+    produce_event_every_ms_ = produce_event_every_ms;
+    upload_events_every_ms_ = upload_events_every_ms;
+    early_stop_event_count_ = early_stop_event_count;
+
+    data_produce_thread_ =
+        std::thread(&UploaderClientCaptureEventCollectorTest::ProduceCaptureData, this);
+    data_upload_thread_ =
+        std::thread(&UploaderClientCaptureEventCollectorTest::UploadCaptureData, this);
+  }
+
+  void Stop() {
+    absl::MutexLock lock{&stop_called_mutex_};
+    stop_called_ = true;
+  }
+
+  void TearDown() override {
+    {
+      absl::MutexLock lock{&stop_called_mutex_};
+      CHECK(stop_called_);
+    }
+
+    CHECK(data_produce_thread_.joinable());
+    data_produce_thread_.join();
+
+    CHECK(data_upload_thread_.joinable());
+    data_upload_thread_.join();
+  }
+
+  UploaderClientCaptureEventCollector collector_;
+  size_t total_uploaded_data_bytes_ = 0;
+
+  absl::Mutex upload_finished_mutex_;
+  bool upload_finished_ ABSL_GUARDED_BY(upload_finished_mutex_) = false;
+  absl::Mutex early_stop_mutex_;
+  bool early_stop_ ABSL_GUARDED_BY(early_stop_mutex_) = false;
+
+ private:
+  // Keep producing capture events until stop is called or `event_count_` events are produced with a
+  // specified produce duration. The last produced event will always be the `CaptureFinishedEvent`.
```

Comment:
Could you update the comment to include the early stop behaviour?

Reply:
Update: Obsolete as we removed the `EarlyStop` test case.